src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -187,96 +187,99 @@ state switch
                         }
                         else
                         {
-                            // Scenario 3.
-                            compilationPair = compilationPair.ReplaceSyntaxTree(oldTree, tree);
-
-                            // The old tree came from some other document with a different ID then we started with -- if the document ID still existed we would have
-                            // been in the Scenario 1 case instead. We'll find the old document ID, remove that state, and then add ours.
-                            var oldDocumentId = DocumentState.GetDocumentIdForTree(oldTree);
-                            Contract.ThrowIfNull(oldDocumentId, $"{nameof(oldTree)} came from the compilation produced by the workspace, so the document ID should have existed.");
-                            inProgressProject = inProgressProject
-                                .RemoveDocuments([oldDocumentId])
-                                .AddDocuments([docState]);
+                            // We're in either scenario 2 or 3. Do we have an existing document to try replacing? Note: the
+                            // file path here corresponds to Document.FilePath. If a document's file path is null, we then
+                            // substitute Document.Name, so we usually expect there to be a unique string regardless.
+                            oldState = inProgressProject.DocumentStates.States.FirstOrDefault(kvp => kvp.Value.FilePath == docState.FilePath).Value;
+                            if (oldState is null)
+                            {
+                                // Scenario 2.
+                                pendingActions = [(inProgressProject, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
+                                inProgressProject = inProgressProject.AddDocuments([docState]);
+                            }
+                            else
+                            {
+                                // Scenario 3. The old doc came from some other document with a different ID then we started
+                                // with -- if the document ID still existed we would have been in the Scenario 1 case
+                                // instead. We'll find the old document ID, remove that state, and then add ours.
+                                var inProgressWithDocumentRemoved = inProgressProject.RemoveDocuments([oldState.Id]);
+                                pendingActions = [
+                                    (inProgressProject, new CompilationAndGeneratorDriverTranslationAction.RemoveDocumentsAction([oldState])),
+                                    (inProgressWithDocumentRemoved, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
+                                inProgressProject = inProgressWithDocumentRemoved
+                                    .AddDocuments([docState]);
```

Comment:
note: i wasn't sure about teh prjects to pass in to pendingActions.  Because tehre are two actions to perform (remove tehn add), it means each action has a different 'old state'.  so, to preserve that, i broke this into individual steps.

Reply:
@jasonmalinowski to check if this is the right way to do things.

Reply:
will do.

Reply:
done in followup pr.